compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1561,71 +1561,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    // `other_struct` itself. See `coerce-in-base-expr.rs`.
                     let fresh_base_ty = self.tcx.mk_adt(*adt, fresh_substs);
                     let base_ty = self.check_expr_has_type_or_error(
                         base_expr,
-                        fresh_base_ty,
+                        self.resolve_vars_if_possible(fresh_base_ty),
                         |_| {
                             error_happened = true;
                         },
                     );
-                    let base_ty = self.shallow_resolve(base_ty);
-                    if let ty::Adt(base_adt, base_substs) = base_ty.kind() && adt == base_adt {
-                        variant
-                            .fields
-                            .iter()
-                            .map(|f| {
-                                let fru_ty = self.normalize_associated_types_in(
-                                    expr_span,
-                                    self.field_ty(base_expr.span, f, base_substs),
-                                );
-                                let ident = self
-                                    .tcx
-                                    .adjust_ident(f.ident(self.tcx), variant.def_id);
-                                if let Some(_) = remaining_fields.remove(&ident) {
-                                    let target_ty =
-                                        self.field_ty(base_expr.span, f, substs);
-                                    let cause = self.misc(base_expr.span);
-                                    match self
-                                        .at(&cause, self.param_env)
-                                        .sup(target_ty, fru_ty)
-                                    {
-                                        Ok(InferOk { obligations, value: () }) => {
-                                            self.register_predicates(obligations)
-                                        }
-                                        // FIXME: Need better diagnostics for `FieldMisMatch` error
-                                        Err(_) => {
-                                            self.report_mismatched_types(
-                                                &cause,
-                                                target_ty,
-                                                fru_ty,
-                                                FieldMisMatch(variant.name, ident.name),
-                                            )
-                                            .emit();
-                                        }
-                                    }
-                                }
-                                self.resolve_vars_if_possible(fru_ty)
-                            })
-                            .collect()
+                    if !error_happened && !base_ty.references_error() {
+                        fru_tys
```

Comment:
why is that needed? What happens if we always use `fru_tys` here

Reply:
yeah, it's way too overprotective i think